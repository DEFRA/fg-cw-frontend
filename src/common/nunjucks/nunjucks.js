import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
