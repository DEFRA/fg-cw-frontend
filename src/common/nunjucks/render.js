import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
