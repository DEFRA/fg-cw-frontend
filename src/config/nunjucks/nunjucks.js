import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
