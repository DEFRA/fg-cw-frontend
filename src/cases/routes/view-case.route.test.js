import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { nunjucks } from "../../common/nunjucks/nunjucks.js";
import { findCaseByIdUseCase } from "../use-cases/find-case-by-id.use-case.js";
import { viewCaseRoute } from "./view-case.route.js";

vi.mock("../use-cases/find-case-by-id.use-case.js");

describe("viewCaseRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(viewCaseRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("returns a list of cases", async () => {
    findCaseByIdUseCase.mockResolvedValue({
      _id: "686baafc9ac48f446c317655",
      caseRef: "PMF-REF-1",
      workflowCode: "pigs-might-fly",
      status: "NEW",
      dateReceived: "2025-03-27T11:34:52Z",
      currentStage: "application-received",
      assignedUser: null,
      payload: {
        clientRef: "PMF-REF-1",
        code: "pigs-might-fly",
        createdAt: "2025-03-27T10:34:52.000Z",
        submittedAt: "2025-03-28T11:30:52.000Z",
        answers: {
          isPigFarmer: true,
          totalPigs: 10,
          whitePigsCount: 2,
          britishLandracePigsCount: 2,
          berkshirePigsCount: 3,
          otherPigsCount: 3,
        },
      },
      stages: [
        {
          id: "application-received",
          taskGroups: [
            {
              id: "review-automated-checks",
              tasks: [
                {
                  id: "review-application-data",
                  status: "pending",
                  title: "Review application data",
                  type: "boolean",
                },
              ],
              title: "Review Automated Checks",
            },
          ],
          title: "Application Received",
          actions: [
            {
              id: "accept",
              label: "Accept",
            },
            {
              id: "reject",
              label: "Reject",
            },
          ],
        },
        {
          id: "assessment",
          taskGroups: [
            {
              id: "check-application",
              tasks: [
                {
                  id: "check-application-and-documents",
                  status: "pending",
                  title: "Check application and documents",
                  type: "boolean",
                },
                {
                  id: "check-find-farm-and-land-payment-data",
                  status: "pending",
                  title: "Check on Find farm and land payment data",
                  type: "boolean",
                },
                {
                  id: "check-rps-dual-funding",
                  status: "pending",
                  title: "Check on RPS (Dual Funding)",
                  type: "boolean",
                },
              ],
              title: "Check Application",
            },
            {
              id: "registration-checks",
              tasks: [
                {
                  id: "confirm-farm-has-cph",
                  status: "pending",
                  title: "Confirm farm has a CPH",
                  type: "boolean",
                },
                {
                  id: "confirm-apha-registration",
                  status: "pending",
                  title: "Confirm APHA registration",
                  type: "boolean",
                },
              ],
              title: "Registration checks",
            },
          ],
          title: "Assessment",
          actions: [
            {
              id: "confirm-approval",
              label: "Confirm Approval",
            },
            {
              id: "confirm-rejection",
              label: "Confirm Rejection",
            },
          ],
        },
        {
          id: "contracted",
          taskGroups: [],
          title: "Contracted",
          actions: [],
        },
      ],
      timeline: [],
      requiredRoles: {
        allOf: ["ROLE_RPA_ADMIN"],
        anyOf: ["ROLE_RPA_ADMIN"],
      },
      pages: {
        details: {
          banner: {
            title: {
              ref: "$.payload.businessName",
              type: "string",
            },
            summary: {
              reference: {
                label: "Reference",
                ref: "$.caseRef",
                type: "string",
              },
              status: {
                label: "Status",
                ref: "$.status",
                type: "string",
              },
              dateReceived: {
                label: "Date Received",
                ref: "$.dateReceived",
                type: "date",
              },
            },
          },
          tabs: {
            caseDetails: {
              title: "Application",
              sections: [
                {
                  title: "Applicant Details",
                  type: "object",
                  component: "list",
                  fields: [
                    {
                      ref: "$.payload.answers.isPigFarmer",
                      type: "boolean",
                      label: "Are you a pig farmer?",
                    },
                  ],
                },
                {
                  title: "Pig Stock Details",
                  type: "object",
                  component: "list",
                  fields: [
                    {
                      ref: "$.payload.answers.totalPigs",
                      type: "number",
                      label: "Total Pigs",
                    },
                    {
                      ref: "$.payload.answers.whitePigsCount",
                      type: "number",
                      label: "How many White pigs do you have?",
                    },
                    {
                      ref: "$.payload.answers.britishLandracePigsCount",
                      type: "number",
                      label: "How many British Landrace pigs do you have?",
                    },
                    {
                      ref: "$.payload.answers.berkshirePigsCount",
                      type: "number",
                      label: "How many Berkshire pigs do you have?",
                    },
                    {
                      ref: "$.payload.answers.otherPigsCount",
                      type: "number",
                      label: "How many Other pigs do you have?",
                    },
                  ],
                },
              ],
            },
          },
        },
      },
    });

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/case-details",
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
