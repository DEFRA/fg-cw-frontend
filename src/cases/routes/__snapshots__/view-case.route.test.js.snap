// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viewCaseRoute > returns a list of cases 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Test application name 1
    </h1>

    <div>
      <div class="app-case-row">
        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">SBI:</strong>
            SBI001
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Reference:</strong>
            frps-private-beta
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Scheme:</strong>
            SFI
            2025
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Date received:</strong>
            11 Jun 2025
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Status:</strong>
            NEW
          </p>
        </div>
      </div>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Case Details</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="#">
                                    
Notes
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>


    <h2 class="govuk-heading-l govuk-!-margin-bottom-1 govuk-!-margin-top-6">
        Case
    </h2>

    <p class="govuk-body">
      submitted:
      <strong>1st October 2023</strong>
    </p>

    <div class="govuk-grid-row">
      <div class="govuk-grid-column-full">
        <div class="govuk-body" data-testid="app-page-body">
            
            

            <table class="govuk-table">
  <tbody class="govuk-table__body">
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">_id</th>
      <td class="govuk-table__cell">68495db5afe2d27b09b2ee47</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">clientRef</th>
      <td class="govuk-table__cell">banana-123</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">businessName</th>
      <td class="govuk-table__cell">Test application name 1</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">code</th>
      <td class="govuk-table__cell">frps-private-beta</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">sbi</th>
      <td class="govuk-table__cell">SBI001</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">scheme</th>
      <td class="govuk-table__cell">SFI</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">dateReceived</th>
      <td class="govuk-table__cell">2025-06-11T10:43:01.603Z</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">submittedAt</th>
      <td class="govuk-table__cell">01/10/2023</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">status</th>
      <td class="govuk-table__cell">NEW</td>
    </tr>
  </tbody>
</table>


            

            <table class="govuk-table">
  <caption class="govuk-table__caption govuk-table__caption--m">Answers</caption>
  <tbody class="govuk-table__body">
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">agreementName</th>
      <td class="govuk-table__cell">Test application name 1</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">scheme</th>
      <td class="govuk-table__cell">SFI</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">year</th>
      <td class="govuk-table__cell">2025</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">hasCheckedLandIsUpToDate</th>
      <td class="govuk-table__cell">true</td>
    </tr>
  </tbody>
</table>


            
              <h2 class="govuk-heading-m">Action cases data</h2>
                <h3 class="govuk-heading-s">
                  Selected land parcel for action 1
                </h3>


                <table class="govuk-table">
  <tbody class="govuk-table__body">
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">parcelId</th>
      <td class="govuk-table__cell">9238</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">sheetId</th>
      <td class="govuk-table__cell">SX0679</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">code</th>
      <td class="govuk-table__cell">CSAM1</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">appliedFor</th>
      <td class="govuk-table__cell">20.23 ha</td>
    </tr>
  </tbody>
</table>

        </div>
      </div>
    </div>
  </div>
        "
`;
